src/code_190B20.c

```diff
@@ -130,23 +284,152 @@ INCLUDE_ASM(s32, "code_190B20", func_8026709C);
-
-INCLUDE_ASM(s32, "code_190B20", create_part_shadow);
-
-INCLUDE_ASM(s32, "code_190B20", remove_part_shadow);
-
-INCLUDE_ASM(s32, "code_190B20", func_802673E4);
-
-INCLUDE_ASM(s32, "code_190B20", remove_player_buffs);
+#ifdef NON_MATCHING
+// Register allocation issues (decorationIndex is placed in s2 for seemingly no reason?).
+// Should be easy to clean up once DecorationTable is more understood
+void add_part_decoration(ActorPart* part, s32 decorationIndex, DecorationId decorationType) {
+    if (part->idleAnimations && (part->flags & 2) == 0) {
+        DecorationTable* decoration = &part->decorationTable->unk_00[decorationIndex];
+        _remove_part_decoration(part, decorationIndex);
+        decoration->decorationType[0] = decorationType;
+        decoration->unk_8BA = 1;
+        decoration->unk_8BC = 0;
+        func_8025CEC8(part);
+    }
+}
+#else
+INCLUDE_ASM(void, "code_190B20", add_part_decoration, ActorPart* part, s32 decorationIndex,
+            DecorationId decorationType);
+#endif
+
+void add_actor_decoration(Actor* actor, s32 decorationIndex, DecorationId decorationType) {
+    ActorPart* part;
+    for (part = actor->partsTable; part != NULL; part = part->nextPart) {
+        if ((part->flags & 0x100001) == 0 && part->idleAnimations && (part->flags & 2) == 0) {
+            add_part_decoration(part, decorationIndex, decorationType);
+        }
+    }
+}
+
+void remove_part_decoration(ActorPart* part, s32 decorationIndex) {
+    _remove_part_decoration(part, decorationIndex);
+}
+
+void remove_actor_decoration(Actor* actor, s32 decorationIndex) {
+    ActorPart* part;
+    for (part = actor->partsTable; part != NULL; part = part->nextPart) {
+        if ((part->flags & 0x100001) == 0 && part->idleAnimations && (part->flags & 2) == 0) {
+            remove_part_decoration(part, decorationIndex);
+        }
+    }
+}
+
+s32 heroes_is_ability_active(Actor* actor, Ability ability) {
+    s32 actorGenus = actor->actorID & 0x700;
+    s32 hasAbility = FALSE;
+
+    if (actorGenus != 0x100) {
+        // Separate ifs required to match
+        if (actorGenus <= 0x100) {
+            if (actorGenus == 0 && (gBattleStatus.flags2 & 0x40) == 0) {
+                hasAbility = is_ability_active(ability);
+            }
+        }
+    } else {
+        hasAbility = is_partner_ability_active(ability);
+    }
+
+    return hasAbility;
+}
+
+void create_part_shadow(s32 actorId, s32 partIndex) {
+    ActorPart* part = get_actor_part(get_actor(actorId), partIndex);
+
+    part->flags &= ~4;
+    part->shadow = create_shadow_type(0, part->currentPos.x, part->currentPos.y, part->currentPos.z);
+    part->shadowScale = part->size[0] / 24.0;
+}
+
+void remove_part_shadow(s32 actorId, s32 partIndex) {
+    ActorPart* part = get_actor_part(get_actor(actorId), partIndex);
+
+    part->flags |= 4;
+    func_80112328(part->shadow);
+}
+
+void create_part_shadow_by_ptr(UNK_TYPE arg0, ActorPart* part) {
```

Comment:
should we use "by reference" rather than "by pointer" for this?

Reply:
Yes. Maybe "by_ref"?